tensorflow/lite/micro/kernels/ceva/fully_connected.cc

```diff
@@ -0,0 +1,257 @@
+#include "tensorflow/lite/kernels/kernel_util.h"
+#include "tensorflow/lite/micro/kernels/ceva/ceva_tflm_lib.h"
+#include "tensorflow/lite/micro/kernels/kernel_util.h"
+//#define MCPS_MEASUREMENT
+#ifdef MCPS_MEASUREMENT
+#include "tensorflow/lite/micro/kernels/ceva/mcps_macros.h"
+#endif
+
+#if defined(CEVA_BX1) || defined(CEVA_SP500)
+extern int32_t* CEVA_TFLM_KERNELS_SCRATCH;
+extern int32_t CEVA_TFLM_KERNELS_SCRATCH_SIZE_VAL;
+#endif  // CEVA platform
+
+namespace tflite {
+namespace {
+
+void* Init(TfLiteContext* context, const char* buffer, size_t length) {
+  TFLITE_DCHECK(context->AllocatePersistentBuffer != nullptr);
+
+  return context->AllocatePersistentBuffer(context,
+                                           sizeof(OpDataFullyConnected));
+}
+
+TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
+  TFLITE_DCHECK(node->user_data != nullptr);
+  TFLITE_DCHECK(node->builtin_data != nullptr);
+
+  auto* data = static_cast<OpDataFullyConnected*>(node->user_data);
+
+  const auto params =
+      static_cast<const TfLiteFullyConnectedParams*>(node->builtin_data);
+
+  const TfLiteTensor* input =
+      GetInput(context, node, kFullyConnectedInputTensor);
+  const TfLiteTensor* filter =
+      GetInput(context, node, kFullyConnectedWeightsTensor);
+  const TfLiteTensor* bias =
+      GetOptionalInputTensor(context, node, kFullyConnectedBiasTensor);
+  TfLiteTensor* output = GetOutput(context, node, kFullyConnectedOutputTensor);
+
+  TF_LITE_ENSURE_TYPES_EQ(context, input->type, output->type);
+  TF_LITE_ENSURE_MSG(context, input->type == filter->type,
+                     "Hybrid models are not supported on TFLite Micro.");
+
+  return CalculateOpDataFullyConnected(context, params->activation, input->type,
+                                       input, filter, bias, output, data);
+}
+
+TfLiteStatus EvalQuantizedInt8CEVA(TfLiteContext* context, TfLiteNode* node,
+                                   const OpDataFullyConnected& data,
+                                   const TfLiteEvalTensor* input,
+                                   const TfLiteEvalTensor* filter,
+                                   const TfLiteEvalTensor* bias,
+                                   TfLiteEvalTensor* output) {
+  tflite::FullyConnectedParams op_params = FullyConnectedParamsQuantized(data);
+
+  int input_shape_DimensionsCount =
+      tflite::micro::GetTensorShape(input).DimensionsCount();
+  int weights_shape_DimensionsCount =
+      tflite::micro::GetTensorShape(filter).DimensionsCount();
+  int* weights_shape_DimsData =
+      const_cast<int*>(tflite::micro::GetTensorShape(filter).DimsData());
+  int bias_shape_DimensionsCount =
+      tflite::micro::GetTensorShape(bias).DimensionsCount();
+  int output_shape_DimensionsCount =
+      tflite::micro::GetTensorShape(output).DimensionsCount();
+  int* output_shape_DimsData =
+      const_cast<int*>(tflite::micro::GetTensorShape(output).DimsData());
+
+  void* params = (void*)&op_params;

```

Comment:
I'm unsure that this cast is needed. If it is, let's directly add to the function call.